fe/fe-core/src/main/java/com/starrocks/rpc/FrontendServiceProxy.java

```diff
@@ -5,26 +5,39 @@
 import com.starrocks.common.ClientPool;
 import com.starrocks.thrift.FrontendService;
 import com.starrocks.thrift.TNetworkAddress;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.thrift.TException;
+import org.apache.thrift.transport.TTransportException;
+
+import java.net.SocketTimeoutException;
 
 public class FrontendServiceProxy {
+    private static final Logger LOG = LogManager.getLogger(FrontendServiceProxy.class);
 
     public static <T> T call(TNetworkAddress address, int timeoutMs, MethodCallable<T> callable) throws Exception {
         FrontendService.Client client = ClientPool.frontendPool.borrowObject(address, timeoutMs);
 
-        boolean returnToPool = false;
-        T res;
+        boolean isConnValid = false;
         try {
-            res = callable.invoke(client);
-            returnToPool = true;
+            T t = callable.invoke(client);
+            isConnValid = true;
+            return t;
+        } catch (TTransportException te) {
+            isConnValid = ClientPool.frontendPool.reopen(client);
+            // do not retry for timeout exception
+            if (!(te.getCause() instanceof SocketTimeoutException) && isConnValid) {
+                return callable.invoke(client);
```

Comment:
I suggest you use a for-loop and a definite constant integer like RETRY_TIMES instead of obscurely setting one retry attempt.

Reply:
done